### DISCUSSION TITLE
Incorrect use of 'subscriptions:' vs. 'subscriptionIds:' in resource discovery configuration in documentation and source code

### DISCUSSION THREAD

**Discussion by cscoon** [tomkerkhove/promitor#1489] [Q&A] [answered]
Throughout the promitor.io website and the GitHub site, there are multiple sample .yaml configurations for resource discovery.  Some of these samples use a node named 'subscriptions' while others name that node 'subscriptionIds'.

When resource discovery is deployed with a configuration file that uses 'subscriptions', the pod goes into a CrashLoopBackOff state, and the logs show the following error:

Promitor Discovery Agent has encountered an unexpected error. Please open an issue at https://github.com/tomkerkhove/promitor/issues to let us know about it.
 System.FormatException: Could not parse the YAML file: (Line: 6, Col: 3, Idx: 101) - (Line: 6, Col: 3, Idx: 101): While scanning a simple key, could not find expected ':'..

## Expected Behavior
The documentation and source code samples for resource discovery configuration should all use valid schema.

## Actual Behavior
The schema used is inconsistent, resulting in deployments that fail on startup.

## Steps to Reproduce the Problem
The following URLs and source code uses the 'subscriptions' rather than 'subscriptionIds' in its sample configuration:

- https://blog.tomkerkhove.be/2021/01/06/announcing-promitor-2-0-bringing-azure-monitor-metrics-where-you-need-them/
- https://promitor.io/configuration/v2.x/resource-discovery
- https://github.com/tomkerkhove/promitor/blob/master/config/promitor/resource-discovery/resource-discovery-declaration.yaml

When a deployment is used with this schema, it fails to start.

## Specifications

  - Version: 2.0.0
  - Platform:
  - Subsystem:

**Comment by tomkerkhove** [comment]
Thanks for reporting!

`Subscriptions` is what should be used as per https://github.com/tomkerkhove/promitor/blob/master/config/promitor/resource-discovery/resource-discovery-declaration.yaml#L4 which is what I'm using.

Would it be possible to share your full declaration and logs please, so that I can have a look?

**Comment by tomkerkhove** [comment]
Also, where did you see that it should be `subscriptionIds`?

**Comment by cscoon** [comment]
I understand that is what that particular documentation claims to be correct, but it contradicts other Promitor documentation, and attempting to deploy with both of those values shows that only subscriptionIds results in a valid deployment.

Here is a deployment declaration using the 'subscriptionIds' and deployed via the helm chart as suggested at [https://promitor.io/deployment/resource-discovery/](url):

(NOTE: the markup is interpreting my yaml dashes as bullet points, apologize... I'm not sure how to escape those)

version: v1
azureLandscape:
  tenantId: <tenant ID guid>
  subscriptionIds:
  - 8b7abd21-5f1f-46ab-9173-ddc5db02e724
  cloud: Global
resourceDiscoveryGroups:
- name: cosmos-db
  type: CosmosDb

The deployment declaration above deploys **correctly and the pod starts**:

<img width="838" alt="subscriptionIds 1" src="https://user-images.githubusercontent.com/40893895/107080349-c4e54280-67b6-11eb-9f59-314821d7d765.PNG">

<img width="757" alt="subscriptionIds 2" src="https://user-images.githubusercontent.com/40893895/107080365-cb73ba00-67b6-11eb-83b7-45ab26d37e7e.PNG">

Here is the same deployment declaration, but using 'subscriptions':

<img width="837" alt="subscriptions 1" src="https://user-images.githubusercontent.com/40893895/107080405-db8b9980-67b6-11eb-95fa-1515b7574ad6.PNG">

<img width="1059" alt="subscriptions 2" src="https://user-images.githubusercontent.com/40893895/107081433-4e494480-67b8-11eb-8b16-b42a187ea468.PNG">

The Promitor charts repo also supports the use of subscriptionIds:

[https://github.com/promitor/charts/blob/main/examples/promitor-agent-resource-discovery.config.yaml](url)

...and subscriptionIds are used in some areas of the promitor.io site as well:

[https://promitor.io/deployment/resource-discovery/](url)

...but this contradicts other places in documentation, which suggest using subscriptions:

[https://promitor.io/configuration/v2.x/resource-discovery.html](url)

**Comment by tomkerkhove** [comment]
Thanks for sharing! So you are deploying through Helm then?

I'll check this asap!

- Tom Kerkhove

________________________________

**Comment by cscoon** [comment]
Update... I'm attempting to run the resource discovery API project locally to debug some issues.  During that, I discovered that when providing the config through a local folder (C:\config), it requires the opposite of what the helm chart requires.

The helm chart requires 'subscriptionIds', but when providing the config locally, the validator requires 'subscriptions'.  So my guess is that the helm chart is loading 'subscriptionIds', and then doing some sort of translation and loading the values into 'subscriptions' for the API runtime.

**Comment by tomkerkhove** [comment, marked as answer]
Yes, that was what I was going to look for.

I'll update the Helm chart, sorry about that!

**Reply by tomkerkhove** [reply]
This is misleading in our Helm chart - I've created https://github.com/promitor/charts/issues/27 for this, sorry for the confusion.

**Reply by cscoon** [reply]
What is the net effect of the commit you made to address this?  Is the net effect that a user who is deploying via the helm chart can now use either 'subscriptions' or 'subscriptionIds' and both will work?

I attempted to deploy Resource Discovery via helm chart with 'subscriptions' in my config file and the pod is still failing to start.

**Reply by tomkerkhove** [reply]
Yes, but `subscriptionIds` is deprecated and removed in v1.0 of the Helm chart (https://github.com/promitor/charts/issues/28)
